src/blog/2025-07-24-oxlint-type-aware.md

```diff
@@ -0,0 +1,164 @@
+now complete in less than 10 seconds.
+
+We achieved this by leveraging [`typescript-go`](https://github.com/microsoft/typescript-go),
+the [10x faster TypeScript](https://devblogs.microsoft.com/typescript/typescript-native-port) written in Go.
+
+## Type-Aware linting
+
+Please refer to
+[Rust-Based JavaScript Linters: Fast, But No Typed Linting Right Now](https://www.joshuakgoldberg.com/blog/rust-based-javascript-linters-fast-but-no-typed-linting-right-now)
+to understand the current status of type-aware linting in the ecosystem.
+
+## Technical Details
+
+The core of this new functionality is [oxc-project/tsgolint](https://github.com/oxc-project/tsgolint).
+
+The tsgolint project was initially prototyped as [typescript-eslint/tsgolint](https://github.com/typescript-eslint/tsgolint).
+However, the `typescript-eslint` team decided not to allocate development resources to this prototype, as they plan to continue their work on `typescript-eslint` for typed linting with ESLint.
+
+@boshen then reached out to @auvred for a forked, scoped-down version adapted for oxlint.
+This version would only contain type-aware rules without the sophisticated configuration resolution a full linter would require.
+
+@auvred generously offered to continue its development under the Oxc organization.
+
+### Architecture
+
+`oxlint` (written in Rust) and `tsgolint` (written in Go) are compiled into their own binaries.
+
+`oxlint` serves as the "frontend" for `tsgolint`, handling the CLI, path traversal, ignore logic, and diagnostic printing.
+
+`tsgolint` acts as the backend for `oxlint`, accepting paths and configuration as input and outputting structured diagnostics.
+
+This creates a simple pipeline:
+
+```
+Oxlint CLI (returns paths + rules + configuration)
+  -> tsgolint (returns diagnostics)
+  -> Oxlint CLI
+```
+
+### `tsgolint`
+
+`tsgolint` does not communicate with typescript-go via public APIs.
+
+Instead, it compiles `typescript-go` by [shimming its internal APIs to make them public](https://github.com/oxc-project/tsgolint/tree/main/shim).
+
+All type-aware rules are written directly against these shimmed APIs.
+
+While this isn't the recommended approach, it works!
+
+## Decision Process
+
+### Write our own type checker
+
+Previous abandoned attempts to implement a type checker included:
+
+- My own attempt at [writing type inference](https://gist.github.com/Boshen/d189de0fe0720a30c5182cb666e3e9a5)
+- [Integrate](https://github.com/oxc-project/oxc/pull/413) [ezno type checker](https://github.com/kaleidawave/ezno) by @kaleidawave
+- [stc](https://github.com/dudykr/stc) by @kdy1
+- (There are also many attempts in the community that did not go far).
+
+Additionally, there's the work-in-progress:
+
+- [Biome 2.0](https://biomejs.dev/blog/biome-v2/) with its own type inference implementation.
+
+We determined that writing our own type inferencer or type checker was not feasible due to budget and time constraints,
```

Comment:
If not elaborate earlier (see comment before), might be worth to mention accuracy too?

Reply:
I suggest removing the reference to "budget". The idea that commercial imperatives and open source ethos inevitably clash and lead to worse outcomes has currency in some quarters. Mentioning financial considerations here might tend to fuel that argument. And I don't think it's fair here - we were struggling with the infeasibility of building a reliable type inference system long before Oxc became part of Void 0.

More generally: Isn't the point here that we decided the surest route to building complete and accurate support for type-aware linting, and keeping up-to-date with TS, is to lean directly on the TS compiler? (following TS-ESLint's example)